### PR TITLE
Pull out a scaled slider class for finer grained control of slice positions

### LINKED ISSUE #50: Slicing through small images does not have the right resolution
### Problem 
When a small image (i.e. one whose extend is only a few mm) is shown with ScalismoUI, slicing does not work properly. With the position slider only few slices can be shown. 
My guess is that this is because the position slider in the ViewportPanel2D takes only integer values, which results in that only slices which are a mm apart are shown. For small images, this is not the right behavior. 

### Expected behaviour:

The number of slices that can be shown should always be large enough to explore the image, irrespective of the size of the image. One possibility would be to always show at least 100 slices (and more if the extend of the sliced scene is larger than 100 mm). 

### Workaround
Using shift-leftclick on the orthogonal planes correctly sets the slicing position. Holding shift clicked and moving the mouse can therefore be used to move navigate through the slices.

### PR DESCRIPTION
**Pull request by Andreas-Forster**
This PR fixes #50 by introducing a new class [SubDividedSliderAdapter](https://github.com/Andreas-Forster/scalismo-ui/blob/098d92cfd6bf82fb3bdbe9c4ccc46412a0a6ecb0/src/main/scala/scalismo/ui/view/util/SubDividedSliderAdapter.scala#L22). The sliders have by default a 100x finer resolution than now. This should fix for most use cases the issue #50 where the slices could not be made visible by using the sliders.

### CODE REVIEW

**Comment by Andreas-Forster** [comment]
Yet, this PR should serve as a discussion basis and testing is still needed before merging it.

File: src/main/scala/scalismo/ui/view/ViewportPanel.scala [hunk 1]
```diff
@@ -121,29 +121,23 @@ class ViewportPanel3D(frame: ScalismoFrame, override val name: String = "3D") ex
 class ViewportPanel2D(frame: ScalismoFrame, val axis: Axis) extends ViewportPanel(frame) {
   override def name: String = axis.toString
 
-  private lazy val positionSlider = new Slider {
+  private val SubDivisions: Double = 100.0
```
Comment (clangguth) [inline src/main/scala/scalismo/ui/view/ViewportPanel.scala]: This is obsolete and can be removed.

File: src/main/scala/scalismo/ui/view/ViewportPanel.scala [hunk 2]
```diff
@@ -121,29 +121,23 @@ class ViewportPanel3D(frame: ScalismoFrame, override val name: String = "3D") ex
...
-  }
+  })
 
   lazy val positionPlusButton = new Button(new Action("+") {
-    override def apply(): Unit = {
-      if (positionSlider.value < positionSlider.max) {
-        positionSlider.value = positionSlider.value + 1
-      }
-    }
+    override def apply(): Unit = positionSlider.up
```
Comment (clangguth) [inline src/main/scala/scalismo/ui/view/ViewportPanel.scala]:
```suggestion
    override def apply(): Unit = positionSlider.up()
```

File: src/main/scala/scalismo/ui/view/ViewportPanel.scala [hunk 3]
```diff
@@ -121,29 +121,23 @@ class ViewportPanel3D(frame: ScalismoFrame, override val name: String = "3D") ex
...
   })
 
   lazy val positionMinusButton = new Button(new Action("-") {
-    override def apply(): Unit = {
-      if (positionSlider.value > positionSlider.min) {
-        positionSlider.value = positionSlider.value - 1
-      }
-    }
+    override def apply(): Unit = positionSlider.down
```
Comment (clangguth) [inline src/main/scala/scalismo/ui/view/ViewportPanel.scala]:
```suggestion
    override def apply(): Unit = positionSlider.down()
```

File: src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala [hunk 4]
```diff
@@ -44,38 +44,30 @@ class SlicingPositionPanel(override val frame: ScalismoFrame) extends BorderPane
...
-    }
+    })
 
     val minus = new Button(new Action("-") {
-      override def apply(): Unit = {
-        if (slider.value > slider.min) {
-          slider.value = slider.value - 1
-        }
-      }
+      override def apply(): Unit = slider.down
```
Comment (clangguth) [inline src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala]:
```suggestion
      override def apply(): Unit = slider.down()
```

File: src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala [hunk 5]
```diff
@@ -44,38 +44,30 @@ class SlicingPositionPanel(override val frame: ScalismoFrame) extends BorderPane
...
     })
 
     val plus = new Button(new Action("+") {
-      override def apply(): Unit = {
-        if (slider.value < slider.max) {
-          slider.value = slider.value + 1
-        }
-      }
+      override def apply(): Unit = slider.up
```
Comment (clangguth) [inline src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala]:
```suggestion
      override def apply(): Unit = slider.up()
```

File: src/main/scala/scalismo/ui/view/util/SubDividedSliderAdapter.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package scalismo.ui.view.util
+
+import scala.swing.Slider
+
+class SubDividedSliderAdapter(private val _slider: Slider, private val factor: Double = 100.0) {
+
+  def slider: Slider = _slider
```
Comment (clangguth) [inline src/main/scala/scalismo/ui/view/util/SubDividedSliderAdapter.scala]: There is no need to introduce a private variable here, as it's an object anyway which is made accessible using this getter.

Therefore, my suggestion is to simply make `slider` a val. In the same vein, there's not really a need to keep `factor` private.

**Review by clangguth — CHANGES_REQUESTED** [review]
Looks good to me in general, I've added a few tiny nitpicking suggestions.

The only thing that still bothers me a bit is that most of the time references are made to the adapter, but sometimes to the contained slider. (listenTo(x.slider.slider)).

This could be overcome by making the adapter itself a ScalismoPublisher that listens to the slider events and re-publishes them on behalf of itself. Then the only occurence where direct access to the contained slider would still be needed is in the layout() calls, because they expect an actual Swing component.

File: src/main/scala/scalismo/ui/view/ViewportPanel.scala (hunk 1 above)
Comment (Andreas-Forster) [reply src/main/scala/scalismo/ui/view/ViewportPanel.scala]: agree - addressed in new commit

File: src/main/scala/scalismo/ui/view/ViewportPanel.scala (hunk 2 above)
Comment (Andreas-Forster) [reply src/main/scala/scalismo/ui/view/ViewportPanel.scala]: agree - addressed in new commit

File: src/main/scala/scalismo/ui/view/ViewportPanel.scala (hunk 3 above)
Comment (Andreas-Forster) [reply src/main/scala/scalismo/ui/view/ViewportPanel.scala]: agree - addressed in new commit

File: src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala (hunk 4 above)
Comment (Andreas-Forster) [reply src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala]: agree - addressed in new commit

File: src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala (hunk 5 above)
Comment (Andreas-Forster) [reply src/main/scala/scalismo/ui/view/properties/SlicingPositionPanel.scala]: agree - addressed in new commit

File: src/main/scala/scalismo/ui/view/util/SubDividedSliderAdapter.scala (hunk 6 above)
Comment (Andreas-Forster) [reply src/main/scala/scalismo/ui/view/util/SubDividedSliderAdapter.scala]: agree - addressed in new commit

**Comment by Andreas-Forster** [comment]
I made the SubDivdedSliderAdapter a ScalismoPublisher. Further, I had to introduce a new FancySliderDecorator as the FancySlider does not handle the position of the labels correctly.

It looks now that we have quite a few sliders classes. Not sure if there is a better solution.

**Comment by clangguth** [comment]
Hmmm... ok, please give me a couple of days to take a closer look at how this can be properly solved.

**Comment by clangguth** [comment]
Sorry, I don't currently find the time for a proper solution, but here's an idea/outline of a "one-size fits all" slider class which covers all existing use cases, maybe someone finds a way to implement it:
- ensure that it works with multiple types, not just integers. This would mean creating a new generic class which mimicks the interface of the existing Slider class, but in a typed way. Implementation-wise it could either wrap a JSlider or a Slider.
- include the formatting part of the FancySlider, so that labels can be shown if needed.
- the tricky part: ensure that it works for any domain (min/max) and that it allows to set a "step size" (or factor, or granularity, or whatever you want to call it).

Essentially, one needs to find functions to map from that "application domain" (AD) to a "slider domain" (SD) and back, where SD only consists of integers and the step size is always 1. However, min/max of SD could be adjusted if needed, but the largest possible range is (Int.MinValue, Int.MaxValue). The mapping between AD and SD would also need to be bijective, so that changes (movement of the slider) properly get propagated back to the application.

**Comment by Andreas-Forster** [comment]
I added a quick draft for a `TypedSlider[A]` class which should work for `Int`, `Float` and `Double`. I replaced all sliders except the anonymous class using directly `scala.swing.Slider` [here](https://github.com/unibas-gravis/scalismo-ui/blob/020aae765941c77042a818f0d8a329df7132a13c/src/main/scala/scalismo/ui/view/properties/GaussianProcessCoefficientsPanel.scala#L59).

A quick test showed that it seems to work. However, I have not yet tested all classes I changed and maybe there is an unhandled corner case. For example, one could require that the step-size is always strictly positive. A lower bound for the step-size based on the min and max value could also make sense, in order to not prevent that the slider's internal value does not overflow. These should however not occur if you use the sliders in a reasonable range. Any opinion if such checks should be added? Do you see other problems?

Further the implicit could probably be removed if one creates a slider class per type (int, float, double). Any opinion here if this should be changed?

**Comment by marcelluethi** [comment]
I lost track about the status of this PR. Is it okay to close it or are we close to something that can be merged?

**Comment by Andreas-Forster** [comment]
I think we can close it for now. We could still pick it up later on.
